src/dmd/optimize.d

```diff
@@ -188,6 +188,52 @@ extern (C++) Expression fromConstInitializer(int result, Expression e1)
     return e;
 }
 
+/* It is possible for constant folding to change an array expression of
+ * unknown length, into one where the length is known.
+ * If the expression 'arr' is a literal, set lengthVar to be its length.
+ */
+package void setLengthVarIfKnown(VarDeclaration lengthVar, Expression arr)
+{
+    if (!lengthVar)
+        return;
+    if (lengthVar._init && !lengthVar._init.isVoidInitializer())
+        return; // we have previously calculated the length
+    size_t len;
+    if (arr.op == TOKstring)
+        len = (cast(StringExp)arr).len;
+    else if (arr.op == TOKarrayliteral)
+        len = (cast(ArrayLiteralExp)arr).elements.dim;
+    else
+    {
+        Type t = arr.type.toBasetype();
+        if (t.ty == Tsarray)
```

Comment:
Looks like this doesn't work.

Reply:
Control flow has likely moved this to the other function.  Should still be kept around though, as it's a utility function.